aspnetcore/fundamentals/primitives/change-tokens.md

```diff
@@ -0,0 +1,192 @@
+
+   A retry is implemented with an exponential back-off. The re-try is present because file locking may occur that temporarily prevents computing a new hash on one of the files.
+
+### Simple startup change token
+
+Register a token consumer `Action` callback for change notifications to the configuration reload token (*Startup.cs*):
+
+[!code-csharp[Main](change-tokens/sample/Startup.cs?name=snippet2)]
+
+`config.GetReloadToken()` provides the token. The callback is the `InvokeChanged` method:
+
+[!code-csharp[Main](change-tokens/sample/Startup.cs?name=snippet3)]
+
+The `state` of the callback is used to pass in the `IHostingEnvironment`. This is useful to determine the correct *appsettings* configuration JSON file to monitor, *appsettings.\<Environment>.json*. File hashes are used to prevent the `WriteConsole` statement from running multiple times due to multiple token callbacks when the configuration file has only actually changed once.
+
+This system runs as long as the app is running and can't be disabled by the user.
+
+### Monitoring configuration changes as a service
+
+In addition to the simple startup token monitoring, the sample also implements monitoring as a service and offers a mechanism to enable and disable the monitoring. First, the sample establishes an `IConfigurationMonitor` interface (*Extensions/ConfigurationMonitor.cs*):
+
+[!code-csharp[Main](change-tokens/sample/Extensions/ConfigurationMonitor.cs?name=snippet1)]
+
+The constructor of the implemented class, `ConfigurationMonitor`, registers a callback for change notifications:
+
+[!code-csharp[Main](change-tokens/sample/Extensions/ConfigurationMonitor.cs?name=snippet2)]
+
+As before, `config.GetReloadToken()` supplies the token. `InvokeChanged` is the callback method. The `state` in this instance is a string that describes the monitoring state. Two properties are used:
+
+* `MonitoringEnabled` indicates if the callback should run its custom code.
+* `CurrentState` describes the current monitoring state for use in the UI.
+
+The `InvokeChanged` method is similar to the earlier approach, except that it:
+
+* Doesn't run its code unless `MonitoringEnabled` is `true`.
+* Sets the `CurrentState` property string to a descriptive message that records the time that the code ran.
+* Notes the current `state` in its `WriteConsole` output.
+
+[!code-csharp[Main](change-tokens/sample/Extensions/ConfigurationMonitor.cs?name=snippet3)]
+
+An instance `ConfigurationMonitor` is registered as a service in `ConfigureServices` of *Startup.cs*:
+
+[!code-csharp[Main](change-tokens/sample/Startup.cs?name=snippet1)]
+
+The Index page offers the user control over configuration monitoring. The instance of `IConfigurationMonitor` is injected into the `IndexModel`:
+
+[!code-csharp[Main](change-tokens/sample/Pages/Index.cshtml.cs?name=snippet1)]
+
+The app responds on form submits when the user selects buttons in the interface to enable and disable monitoring:
```

Comment:
A button enables/disables monitoring.

No need to provide all those other details.